src/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.Http2.cs

```diff
@@ -0,0 +1,253 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.IO;
+using System.Net.Http.Headers;
+using System.Net.Sockets;
+using System.Net.Test.Common;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Xunit;
+using Xunit.Abstractions;
+
+namespace System.Net.Http.Functional.Tests
+{
+    using Configuration = System.Net.Test.Common.Configuration;
+
+    public sealed class SocketsHttpHandler_HttpClientHandler_Http2_Test : HttpClientTestBase
+    {
+        protected override bool UseSocketsHttpHandler => true;
+
+        [Fact]
+        public async Task Http2_ClientPreface_Sent()
+        {
+            HttpClientHandler handler = CreateHttpClientHandler();
+            handler.ServerCertificateCustomValidationCallback = TestHelper.AllowAllCertificates;
+            TestHelper.EnsureHttp2Feature(handler);
+
+            using (var server = Http2LoopbackServer.CreateServer())
+            using (var client = new HttpClient(handler))
+            {
+                Task sendTask = client.GetAsync(server.Address);
+
+                string connectionPreface = await server.AcceptConnectionAsync();
+
+                Assert.Equal(connectionPreface, "PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n");
```

Comment:
Always arrange Assert.Equal() parameters as EXPECTED first and then ACTUAL.

Reply:
That is why the test failures are showing weird results:

```
Assert.Equal() Failure
          ↓ (pos 0)
Expected: GET / HTTP\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0\\0
Actual:   PRI * HTTP/2.0\\r\
\\r\
SM\\r\
\\r\

          ↑ (pos 0)
```

Reply:
The failures are really that you EXPECT to get "PRI * etc." and are actually getting "GET / HTTP" etc.

The OS's that are failing are probably due to ALPN not working right.  We know that OSX has problems with ALPN and so you'll need to disable the tests for those platforms (there is already an active issue for implementing ALPN on MAC). The other OS platform failures I'm not sure about.

Reply:
I'll see if I can set up VMs for some of the affected OS to investigate what's going on here. I should be able to at a minimum see if h2c is being listed via ALPN.

Thanks for pointing out that the asserts were arranged incorrectly. I'll fix that.

Reply:
I've updated this PR to only run the tests when ALPN is supported, and to fix the order of the Assert.Equal arguments.

Reply:
>and to fix the order of the Assert.Equal arguments

I'm still seeing the wrong order:

```c#
Assert.Equal(connectionPreface, "PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n"); 
```

Reply:
Sorry, this is what I get for working on two machines. Fixed, and checked over the others.